File: control-plane/sdk/python/aosdk/mcp.py
"""MCP client bootstrap (issue #41).

A small, typed bootstrap client for the platform's tenant-scoped MCP tool
gateway (issue #20): it speaks minimal JSON-RPC 2.0 over the injected
transport and exposes the lifecycle a governed consumer needs —

- :meth:`McpClient.initialize` — protocol negotiation
  (``protocolVersion: 2024-11-05``) + client info;
- :meth:`McpClient.ping` — liveness;
- :meth:`McpClient.list_tools` — the declared tool catalog (narrowed to the
  session's allowed tools by the platform);
- :meth:`McpClient.call_tool` — one ``tools/call`` carrying the tenant-scoped
  session token in ``context.session`` (the platform enforces identity,
  allowlist, rate and audit — issue #20).

The client never asserts a tenant of its own choosing: ``context.tenantId``
is derived from the verified session claims so a caller cannot cross tenants.
"""

from __future__ import annotations

import uuid
from typing import Any, Dict, List, Mapping, Optional

from .auth import SessionToken, TokenSource, verify_not_expired
from .errors import ConfigurationError, McpError
from .model import ToolDefinition, ToolResult

#: Protocol version the gateway advertises (issue #20).
PROTOCOL_VERSION = "2024-11-05"
#: Default MCP endpoint path (a deployment configures its own gateway URL).
DEFAULT_MCP_PATH = "/v1/mcp"


class McpClient:
    """JSON-RPC 2.0 MCP client bootstrap (typed, transport-injected)."""

    def __init__(
        self,
        transport: Any,
        *,
        token_source: Optional[TokenSource] = None,
        tenant_id: Optional[str] = None,
        path: str = DEFAULT_MCP_PATH,
        client_name: str = "agent-orchestrator-sdk",
        client_version: str = "0.1.0",
    ) -> None:
        self._transport = transport
        self._token_source = token_source or TokenSource()
        self._tenant_id = tenant_id
        self._path = path
        self._client_name = client_name
        self._client_version = client_version
        self._next_id = 1

    # -- auth ---------------------------------------------------------------- #
    def _session(self) -> SessionToken:
        token = self._token_source.require()
        session = verify_not_expired(token)  # fail closed on a lapsed token
        return session

    # -- JSON-RPC plumbing --------------------------------------------------- #
    def _rpc(self, method: str, params: Mapping[str, Any]) -> Dict[str, Any]:
        token = self._token_source.require()
        request_id = self._next_id
        self._next_id += 1
        payload: Dict[str, Any] = {
            "jsonrpc": "2.0",
            "id": request_id,
            "method": method,
            "params": dict(params),
        }
        response = self._transport.request("POST", self._path, body=payload, token=token)
        if not isinstance(response, Mapping):
            raise McpError(-32700, "response was not a JSON-RPC object")
        error = response.get("error")
        if error is not None:
            raise McpError(
                int(error.get("code") or -32603),
                str(error.get("message") or "json-rpc error"),
                error.get("data"),
            )
        if "result" not in response:
            raise McpError(-32603, "json-rpc response had no result")
        return response["result"]

    # -- lifecycle ------------------------------------------------------------ #
    def initialize(self) -> Dict[str, Any]:
        """Negotiate the protocol and advertise client capabilities."""
        return self._rpc(
            "initialize",
            {
                "protocolVersion": PROTOCOL_VERSION,
                "capabilities": {"tools": {}},
                "clientInfo": {"name": self._client_name, "version": self._client_version},
            },
        )

    def ping(self) -> bool:
        """Return ``True`` when the gateway answers ``ping``."""
        result = self._rpc("ping", {})
        return result is not None

    def list_tools(self) -> List[ToolDefinition]:
        """The declared tool catalog (platform-narrowed to the session)."""
        self._session()  # fail closed: a tools/list for an absent/lapsed session
        result = self._rpc(
            "tools/list",
            {"context": {"session": self._raw_token()}},
        )
        return [ToolDefinition.from_dict(item) for item in (result.get("tools") or [])]

    def call_tool(
        self,
        name: str,
        arguments: Optional[Mapping[str, Any]] = None,
    ) -> ToolResult:
        """Call one declared tool with ``context.session`` + ``context.tenantId``.

        ``context.tenantId`` is derived from the verified session claims (the
        platform refuses an explicit tenant that disagrees — no cross-tenant).
        """
        session = self._session()
        tenant_id = self._tenant_id or session.tenant_id
        if not tenant_id:
            raise ConfigurationError("session token must carry a tenantId claim")
        result = self._rpc(
            "tools/call",
            {
                "name": name,
                "arguments": dict(arguments or {}),
                "context": {"session": self._raw_token(), "tenantId": tenant_id},
            },
        )
        return ToolResult.from_dict(result)

    # -- helpers -------------------------------------------------------------- #
    def _raw_token(self) -> str:
        return self._token_source.require()


def new_request_id() -> str:
    """Opaque JSON-RPC request id (uuid hex) for callers that build messages."""
    return uuid.uuid4().hex

File: control-plane/sdk/typescript/src/mcp.ts
/**
 * MCP client bootstrap (TypeScript) — issue #41 AC1.
 *
 * A small, typed bootstrap client for the platform's tenant-scoped MCP tool
 * gateway (issue #20): minimal JSON-RPC 2.0 over the injected transport with
 * `initialize` (protocol `2024-11-05`), `ping`, `tools/list` and `tools/call`.
 *
 * The client never asserts a tenant of its own choosing: `context.tenantId`
 * is derived from the verified session claims so a caller cannot cross
 * tenants, and `context.session` rides on every `tools/call` (the platform
 * enforces identity, allowlist, rate and audit — issue #20).
 */

import { SessionToken, TokenSource, verifyNotExpired } from "./auth.js";
import { ConfigurationError, McpError } from "./errors.js";
import type { Transport } from "./transport.js";
import { ToolDefinition, ToolResult } from "./model.js";

/** Protocol version the gateway advertises (issue #20). */
export const PROTOCOL_VERSION = "2024-11-05";
/** Default MCP endpoint path (a deployment configures its own gateway URL). */
export const DEFAULT_MCP_PATH = "/v1/mcp";

export interface McpClientOptions {
  tokenSource?: TokenSource;
  tenantId?: string;
  path?: string;
  clientName?: string;
  clientVersion?: string;
}

interface JsonRpcEnvelope {
  jsonrpc: string;
  id: number;
  result?: unknown;
  error?: { code?: number; message?: string; data?: Record<string, unknown> };
}

export class McpClient {
  private readonly transport: Transport;
  private readonly tokenSource: TokenSource;
  private readonly tenantId?: string;
  private readonly path: string;
  private readonly clientName: string;
  private readonly clientVersion: string;
  private nextId = 1;

  constructor(transport: Transport, options: McpClientOptions = {}) {
    this.transport = transport;
    this.tokenSource = options.tokenSource ?? new TokenSource();
    this.tenantId = options.tenantId;
    this.path = options.path ?? DEFAULT_MCP_PATH;
    this.clientName = options.clientName ?? "agent-orchestrator-sdk";
    this.clientVersion = options.clientVersion ?? "0.1.0";
  }

  /** Fail closed: the current session must be present and not expired. */
  private session(): SessionToken {
    const token = this.tokenSource.require();
    return verifyNotExpired(token);
  }

  private rawToken(): string {
    return this.tokenSource.require();
  }

  /** One JSON-RPC round trip (throws `McpError` on a JSON-RPC error reply). */
  private async rpc(method: string, params: Record<string, unknown>): Promise<unknown> {
    const id = this.nextId++;
    const payload: JsonRpcEnvelope = {
      jsonrpc: "2.0",
      id,
      method,
      params,
    };
    const response = (await this.transport.request("POST", this.path, {
      body: payload,
      token: this.rawToken(),
    })) as unknown as JsonRpcEnvelope;
    if (typeof response !== "object" || response === null || !("id" in response)) {
      throw new McpError(-32700, "response was not a JSON-RPC object");
    }
    if (response.error !== undefined && response.error !== null) {
      throw new McpError(
        response.error.code ?? -32603,
        response.error.message ?? "json-rpc error",
        response.error.data
      );
    }
    if (!("result" in response)) {
      throw new McpError(-32603, "json-rpc response had no result");
    }
    return response.result;
  }

  /** Negotiate the protocol and advertise client capabilities. */
  async initialize(): Promise<Record<string, unknown>> {
    const result = await this.rpc("initialize", {
      protocolVersion: PROTOCOL_VERSION,
      capabilities: { tools: {} },
      clientInfo: { name: this.clientName, version: this.clientVersion },
    });
    return result as Record<string, unknown>;
  }

  /** Return `true` when the gateway answers `ping`. */
  async ping(): Promise<boolean> {
    const result = await this.rpc("ping", {});
    return result !== undefined;
  }

  /** The declared tool catalog (platform-narrowed to the session). */
  async listTools(): Promise<ToolDefinition[]> {
    this.session(); // fail closed on an absent/lapsed session
    const result = (await this.rpc("tools/list", {
      context: { session: this.rawToken() },
    })) as { tools?: ToolDefinition[] };
    return result.tools ?? [];
  }

  /** Call one declared tool with `context.session` + `context.tenantId`. */
  async callTool(name: string, arguments_: Record<string, unknown> = {}): Promise<ToolResult> {
    const session = this.session();
    const tenantId = this.tenantId ?? session.tenantId;
    if (!tenantId) {
      throw new ConfigurationError("session token must carry a tenantId claim");
    }
    const result = await this.rpc("tools/call", {
      name,
      arguments: arguments_,
      context: { session: this.rawToken(), tenantId },
    });
    return result as ToolResult;
  }
}
